### ISSUE TITLE
Server side rendering

### ISSUE THREAD

**Issue by jonbjk** [systemjs/systemjs#2273]
Hi
I have been experimenting with server side rendering. On first request it renders fine, but on repeated requests it fails. On repeat requests the "shell" entry in the importMap is loaded, but it seems the "react" entry is not.

````
import express from "express";
import favicon from "serve-favicon";
import { System, applyImportMap } from "systemjs";
import ReactDOMServer from "react-dom/server";

const port = 3000;

const reactVersion = ReactDOMServer.version;
const systemJsVersion = require(__dirname +
  "/node_modules/systemjs/package.json").version;

// Fake the Bundle Registry call
const getImportMap = () =>
  Promise.resolve({
    imports: {
      "bootstrap-client": "http://localhost:5000/bootstrap-client.js",
      shell: "http://localhost:5000/shell.js",
      react: `https://unpkg.com/react@${reactVersion}/umd/react.development.js`,
      "react-dom": `https://unpkg.com/react-dom@${reactVersion}/umd/react-dom.development.js`,
    },
  });

const app = express();

app
  .use(favicon(__dirname + "/public/favicon.ico"))

  .get("/*", async (req, res) => {
    const importMap = await getImportMap();

    const systemLoader = new System.constructor();
    applyImportMap(systemLoader, importMap);
    const markup = await systemLoader
      .import("shell")
      .then((m) => ReactDOMServer.renderToString(m.default()));

    res.status(200).send(
      `<!doctype html>
    <html lang="">
    <head>
     </head>
    <body>
        <div id="root">${markup}</div>
        <script type="systemjs-importmap">
            ${JSON.stringify(importMap)}
        </script>
        <script>
        const systemJsLoaderTag = document.createElement("script");
        systemJsLoaderTag.src = "https://unpkg.com/systemjs@${systemJsVersion}/dist/system.min.js";
        systemJsLoaderTag.addEventListener("load", function () {
            System.import("bootstrap-client").then(
                function (module) {
                    module.default();
                }
            )
        });
        document.head.appendChild(systemJsLoaderTag);
    </script>     
    </body>
    </html>`
    );
  })
  .listen(port, function (err) {
    if (err) {
      console.error(err);
      return;
    }
    console.log(`> Started on port ${port}`);
  });
````

I thought that 
````
    const systemLoader = new System.constructor();
````
would give me a fresh instance on each request, but I have a feeling that this is not the case.

Can anyone see/explain what I am missing?

**Comment by joeldenning** [comment]
Hi @jonbjk, thanks for sharing your code here.

> On repeat requests the "shell" entry in the importMap is loaded, but it seems the "react" entry is not.

Could you clarify what you mean here? Is `react` failing to load entirely? Do you get an error message?

> I thought that `new System.constructor()` would give me a fresh instance on each request, but I have a feeling that this is not the case.

Yes, this should be the case.

For whatever it's worth, i also have been working on SSR with import maps, and ended up creating https://github.com/node-loader/node-loader-import-maps which allows you to use import maps with the built-in NodeJS module resolution. More details at https://single-spa.js.org/docs/ssr-overview

**Comment by jonbjk** [comment]
Hi @joeldenning, thanks for the prompt reply.
Even if I test with the simplest "hello world" React component that look like:
````
System.register(['react'], function (exports) {
	'use strict';
	var React;
	return {
		setters: [function (module) {
			React = module.default;
		}],
		execute: function () {

			var MyComp = exports('default', function MyComp() {
			  return /*#__PURE__*/React.createElement("div", null, "Hello shell");
			});

		}
	};
});
````
On first request it renders fine both server and client side.
On second request I get the following error from node:
````
(node:3854) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'createElement' of undefined
````
Indicating that React has not loaded on second request. I do not really understand why though if System.constructor() gives me a fresh new instance.

I will give your node loader a try tomorrow.

**Comment by joeldenning** [comment]
I don't see anything wrong with your code - could you provide a demonstration of the issue with a github repo?

**Comment by jonbjk** [comment]
Solved it. 
There were two problems with the code above.
1. Loading React and ReactDOM as UMD sets them and a ref to System on global. This was the reason that second request was corrupted.Using SystemJs versions from https://github.com/esm-bundle solved that problem and the simple test component was rendering fine on server and client.

2. However, when I tried to render a more normal component with some hooks, I got errors because there were two instances of React in play. ReactDOMServer has React as peer dependency, so there is one instance from the server and another pulled in via SystemJs. So I had to compile a SystemJs version of ReactDOMServer to solve the problem.

Now server side rendering of dynamic loaded modules seem to work. I have not tried more complex components yet, but I think this is a workable start.
